File: compy_cli/test_dir/cpp/src/dataclasses_test/with_methods.cpp
#include "dataclasses_test/with_methods.h"
#include "py_str.h"
#include "pypp_util/print.h"

int DataClassWithMethods::add(int x) { return field1 + x; }

int DataClassWithMethods::subtract(int x) { return field1 - x; }

struct _PrivateDataClassWithMethods {
    int field1;
    _PrivateDataClassWithMethods(int a_field1) : field1(std::move(a_field1)) {}
    int add(int x) { return field1 + x; }
};

void dataclass_with_methods_fn() {
    print(PyStr("DATACLASS WITH METHODS RESULTS:"));
    DataClassWithMethods k = DataClassWithMethods(2);
    print(k.add(1), k.subtract(1));
    _PrivateDataClassWithMethods p = _PrivateDataClassWithMethods(3);
    print(p.add(2));
}

File: pypp_cli/test_dir/src/dataclasses_test/with_methods.py
from pypp_python import dataclass


@dataclass
class DataClassWithMethods:
    field1: int

    def add(self, x: int) -> int:
        return self.field1 + x

    def subtract(self, x: int) -> int:
        return self.field1 - x


@dataclass
class _PrivateDataClassWithMethods:
    field1: int

    def add(self, x: int) -> int:
        return self.field1 + x


def dataclass_with_methods_fn():
    print("DATACLASS WITH METHODS RESULTS:")
    # method signatures in .h file, implementations in .cpp file
    k: DataClassWithMethods = DataClassWithMethods(2)
    print(k.add(1), k.subtract(1))
    # private dataclass
    p: _PrivateDataClassWithMethods = _PrivateDataClassWithMethods(3)
    print(p.add(2))
